File: algorithms/leetcode/301-remove-invalid-parentheses/RemoveInvalidParentheses.java
class Solution {
    private void dfs(String s, int pos, StringBuilder buffer, int freeL, int freeR, int opened, Set<String> result) {
        if (freeL < 0 || freeR < 0 || opened < 0 || pos > s.length()) {
            return;
        }
        
        if (pos == s.length()) {
            if (freeL == 0 && freeR == 0 && opened == 0) {
                result.add(buffer.toString());
            }
            
            return;
        }
        
        char ch = s.charAt(pos);

        if (ch == '(') {
            dfs(s, pos + 1, buffer, freeL - 1, freeR, opened, result);
            buffer.append(ch);
            dfs(s, pos + 1, buffer, freeL, freeR, opened + 1, result);
        } else if (ch == ')') {
            dfs(s, pos + 1, buffer, freeL, freeR - 1, opened, result);
            buffer.append(ch);
            dfs(s, pos + 1, buffer, freeL, freeR, opened - 1, result);
        } else {
            buffer.append(ch);
            dfs(s, pos + 1, buffer, freeL, freeR, opened, result);
        }
        
        buffer.setLength(buffer.length() - 1);
    }
    
    public List<String> removeInvalidParentheses(String s) {
        int freeL = 0;
        int freeR = 0;
        
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            
            if (ch == '(') {
                freeL++;
            } else if (ch == ')') {
                if (freeL > 0) {
                    freeL--;
                } else {
                    freeR++;
                }
            }
        }
        
        StringBuilder buffer = new StringBuilder();
        Set<String> result = new HashSet<>();
        dfs(s, 0, buffer, freeL, freeR, 0, result);
        
        return new LinkedList<String>(result);
    }
}

File: leetcode/301-remove-invalid-parentheses/remove-invalid-parentheses.js
/**
 * @param {string} s
 * @return {string[]}
 */
const removeInvalidParentheses = function(s) {
  const results = new Set();

  const buffer = [];

  let freeL = 0;
  let freeR = 0;

  for (const ch of s) {
    if (ch == '(') {
      freeL++;
    } else if (ch == ')') {
      if (freeL > 0) {
        freeL--;
      } else {
        freeR++;
      }
    }
  }

  const dfs = (pos, curFreeL, curFreeR, opened) => {
    if (curFreeL < 0 || curFreeR < 0 || opened < 0) {
      return;
    }

    if (pos === s.length) {
      if (curFreeL === 0 && curFreeR === 0 && opened === 0) {
        results.add(buffer.join(''));
      }

      return;
    }

    const len = buffer.length;
    const ch = s[pos];

    if (ch === '(') {
      dfs(pos + 1, curFreeL - 1, curFreeR, opened);
      buffer.push(ch);
      dfs(pos + 1, curFreeL, curFreeR, opened + 1);
    } else if (ch === ')') {
      dfs(pos + 1, curFreeL, curFreeR - 1, opened);
      buffer.push(ch);
      dfs(pos + 1, curFreeL, curFreeR, opened - 1);
    } else {
      buffer.push(ch);
      dfs(pos + 1, curFreeL, curFreeR, opened);
    }

    buffer.length = len;
  };

  dfs(0, freeL, freeR, 0);

  return [...results];
};
